src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -86,5 +116,256 @@ private protected IReadOnlyList<PackageReference> GetTransitivePackageReferences
+        }
+
+        /// <summary>
+        /// Obtains cached entry for a given transitive package
+        /// </summary>
+        /// <param name="transitivePackage">Identity of transitive package</param>
+        /// <returns>A <see cref="TransitiveEntry"/> object, or <c>null</c> if not found</returns>
+        /// <seealso cref="ClearCachedTransitiveOrigins"/>
+        /// <seealso cref="SetCachedTransitiveOrigin(PackageIdentity, TransitiveEntry)"/>
+        internal TransitiveEntry GetCachedTransitiveOrigin(PackageIdentity transitivePackage)
+        {
+            string key = GetTransitiveCacheKey(transitivePackage);
+
+            if (TransitiveOriginsCache.ContainsKey(key))
+            {
+                return TransitiveOriginsCache[key];
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Replaces cached entry for a given transitive package with a <see cref="TransitiveEntry"/>
+        /// </summary>
+        /// <param name="transitivePackage">Identity of transitive package</param>
+        /// <param name="origins">Packages identified as package origins</param>
+        /// <seealso cref="ClearCachedTransitiveOrigins"/>
+        /// <seealso cref="GetCachedTransitiveOrigin(PackageIdentity)"/>
+        internal void SetCachedTransitiveOrigin(PackageIdentity transitivePackage, TransitiveEntry origins)
+        {
+            string key = GetTransitiveCacheKey(transitivePackage);
+            TransitiveOriginsCache[key] = origins;
+        }
+
+        /// <summary>
+        /// Clears Transitive Origins cache
+        /// </summary>
+        /// <seealso cref="GetCachedTransitiveOrigin(PackageIdentity)"/>
+        /// <seealso cref="SetCachedTransitiveOrigin(PackageIdentity, TransitiveEntry)"/>
+        internal void ClearCachedTransitiveOrigins()
+        {
+            TransitiveOriginsCache.Clear();
+        }
+
+        /// <summary>
+        /// Obtains <see cref="PackageSpec"/> object from assets file from disk
+        /// </summary>
+        /// <param name="ct">Cancellation token</param>
+        /// <remarks>Each project implementation has its own way for gathering assets file</remarks>
+        /// <returns>A <see cref="PackageSpec"/> filled from assets file on disk</returns>
+        internal abstract ValueTask<PackageSpec> GetPackageSpecAsync(CancellationToken ct);
+
+        /// <summary>
+        /// Decides wether cached <see cref="PackageSpec"/> differs from assets file on disk
+        /// </summary>
+        /// <param name="actual">A <see cref="PackageSpec"/> read from disk</param>
+        /// <param name="cached">Cached <see cref="PackageSpec"/></param>
+        /// <returns><c>true</c> if current <see cref="PackageSpec"/> differs from cached objects</returns>
+        internal abstract bool IsPackageSpecDifferent(PackageSpec actual, PackageSpec cached);
```

Comment:
See related comments. Not sure this needs to be a method.

Reply:
Will be addressed in https://github.com/NuGet/Home/issues/11397